Update GDScript Export Docs (#6481)
godotengine/godot-docs: tutorials/scripting/gdscript/gdscript_exports.rst

In Godot, class members can be exported. This means their value gets saved along
with the resource (such as the :ref:`scene <class_PackedScene>`) they're
attached to. They will also be available for editing in the property editor.
Exporting is done by using the ``@export`` annotation.

::

    @export var number: int = 5

> In that example the value `5` will be saved and visible in the property editor.

Reviewer: This probably needs two ticks on each side, I think it should be formatted as code.

can modify values that later influence how the program runs. For this, a
special export syntax is provided.

> For more information about export annotations see :ref:`GDScript annotations <class_@GDScript>`.

Reviewer: I agree that the original sentence is no longer valid, but that's a weird thing to say in an article that exists to provide more information about export annotations itself :)

::

    @export var resource: Resource
    @export var node: Node

Grouping Exports
----------------

It is possible to group your exported properties inside the Inspector
with the :ref:`@export_group <class_@GDScript_annotation_@export_group>`
annotation. Every exported property after this annotation will be added to
the group. Start a new group or use ``@export_group("")`` to break out.

::

    @export_group("My Properties")
    @export var number = 3

The second argument of the annotation can be used to only group properties
with the specified prefix.

Groups cannot be nested, use :ref:`@export_subgroup <class_@GDScript_annotation_@export_subgroup>`
to create subgroups within a group.

::

    @export_subgroup("Extra Properties")
    @export var string = ""
    @export var flag = false

You can also change the name of your main category, or create additional
categories in the property list with the :ref:`@export_category <class_@GDScript_annotation_@export_category>`
annotation.

::

    @export_category("Main Category")
    @export var number = 3
    @export var string = ""
    
    @export_category("Extra Category")
    @export var flag = false

> .. note::
>     The list of properties is organized based on the class inheritance and

Reviewer: A new line is required here.

```suggestion
.. note::

    The list of properties is organized based on the class inheritance and
```

    @export var number: int

Resources and nodes can be exported.

::

    @export var resource: Resource
    @export var node: Node

Grouping Exports
----------------

> It is possible to group your exported variables inside the Inspector with the :ref:`@export_group <class_@GDScript_annotation_@export_group>` annotation.
> Groups cannot be nested, use :ref:`@export_subgroup <class_@GDScript_annotation_@export_subgroup>` to add subgroups within groups.

Reviewer: These lines should be wrapped to keep the style of the document consistent. We don't do it everywhere, but in this particular article we do, so the formatting style should be preserved.

```suggestion
It is possible to group your exported properties inside the Inspector
with the :ref:`@export_group <class_@GDScript_annotation_@export_group>`
annotation. Every exported property after this annotation will be added to
the group. Start a new group or use ``@export_group("")`` to break out.

::

    @export_group("My Properties")
    @export var number = 3
    @export var string = ""

Second argument of the annotation can be used to only group properties
with the specified prefix.

Groups cannot be nested, use :ref:`@export_subgroup <class_@GDScript_annotation_@export_subgroup>`
to create subgroups within a group.
```

::

>     @export_group("My Properties")
>     @export var number = 3
>     @export var string = ""

Reviewer: ```suggestion
    @export_group("My Properties")
    @export var number = 3
    
    @export_subgroup("Extra Properties")
    @export var string = ""
    @export var flag = false
```

> You can also change the name of the category with the :ref:`@export_category <class_@GDScript_annotation_@export_category>` annotation.

Reviewer: ```suggestion
You can also change the name of your main category, or create additional
categories in the property list with the :ref:`@export_category <class_@GDScript_annotation_@export_category>`
annotation.

.. note::

    The list of properties is organized based on the class inheritance and
    new categories break that expectation. Use them carefully, especially
    when creating projects for public use.
```

::

>     @export_category("My Category")

Reviewer: ```suggestion
    @export_category("Main Category")
    @export var number = 3
    @export var string = ""
    
    @export_category("Extra Category")
    @export var flag = false
```

Using bit flags requires some understanding of bitwise operations.
If in doubt, use boolean variables instead.

Exporting enums
---------------

> The Editor will enumerate as ``0`` THING_1, ``1`` THING_2, ``-1`` ANOTHER_THING.

Reviewer: ```suggestion
Properties can be exported with a type hint referencing an enum to limit their values
to the values of the enumeration. The editor will create a widget in the Inspector, enumerating
the following as THING_1, THING_2, ANOTHER_THING. The value will be stored as an integer.
```

::

    enum NamedEnum {THING_1, THING_2, ANOTHER_THING = -1}
    @export var x: NamedEnum

> To create an export specific enum you can use the ``@export_enum`` annotation.
> 
> The Editor will enumerate as ``0`` Warrior, ``1`` Magician, ``2`` Thief.

Reviewer: ```suggestion
Integer and string properties can also be limited to a specific list of values using
the :ref:`@export_enum <class_@GDScript_annotation_@export_enum>` annotation.
The editor will create a widget in the Inspector, enumerating the following as Warrior,
Magician, Thief. The value will be stored as an integer, corresponding to the index
of the selected option (i.e. ``0``, ``1``,  or ``2``).
```

::

    @export_enum("Warrior", "Magician", "Thief") var character_class

> If the type is a String, the editor will enumerate with string names.

Reviewer: ```suggestion
If the type is String, the value will be stored as a string.
```
